components/swim/src/server/inbound.rs

```diff
@@ -0,0 +1,171 @@
+            tx_outbound: tx_outbound
+        }
+    }
+
+    pub fn run(&self) {
+        let mut recv_buffer: Vec<u8> = vec![0; 1024];
+        loop {
+            if self.server.pause.load(Ordering::Relaxed) {
+                thread::sleep(Duration::from_millis(100));
+                continue;
+            }
+            match self.server.socket.recv_from(&mut recv_buffer[..]) {
+                Ok((length, addr)) => {
+                    if self.server.check_blacklist(&addr) {
+                        println!("Not processing message from {} - it is blacklisted", addr);
+                        continue;
+                    }
+                    let msg: Swim = match protobuf::parse_from_bytes(&recv_buffer[0..length]) {
+                        Ok(msg) => msg,
+                        Err(e) => {
+                            error!("Error parsing protobuf: {:?}", e);
+                            continue;
+                        }
+                    };
+                    debug!("SWIM Message: {:?}", msg);
+                    match msg.get_field_type() {
+                        Swim_Type::PING => {
+                            self.process_ping(addr, msg);
+                        },
+                        Swim_Type::ACK => {
+                            self.process_ack(addr, msg);
+                        }
+                        Swim_Type::PINGREQ => {
+                            self.process_pingreq(addr, msg);
+                        }
+                    }
+                },
+                Err(e) => {
+                    match e.raw_os_error() {
+                        Some(35) => {
+                            // This is the normal non-blocking result
+                        },
+                        Some(_) => {
+                            error!("UDP Receive error: {}", e);
+                            debug!("UDP Receive error debug: {:?}", e);
+                        },
+                        None => {
+                            error!("UDP Receive error: {}", e);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn process_pingreq(&self, addr: SocketAddr, msg: Swim) {
+        trace_swim!(&self.server, "recv-pingreq", &format!("{}", addr), Some(&msg));
+        let target_addr = {
+            let ml = match self.server.member_list.read() {
+                Ok(ml) => ml,
+                Err(e) => panic!("The member list lock is poisoned: {:?}", e)
+            };
+            match ml.get(msg.get_pingreq().get_target().get_id()) {
+                Some(member) => {
+                    member.socket_address()
+                },
+                None => {
+                    error!("PingReq request {:?} for invalid target", msg);
+                    return;
+                }
+            }
+        };
+        outbound::ping(self.server, target_addr, Some(Member::new_from_proto(msg.get_pingreq().get_from().clone())));
```

Comment:
You can take ownership instead of cloning with `msg.take_pingreq().take_from()` here